Modules/_abc.c

```diff
@@ -0,0 +1,853 @@
+    int r = PyObject_RichCompareBool(impl->_abc_negative_cache_version,
+                                     abc_invalidation_counter, Py_LT);
+    assert(r >= 0);  // Both should be PyLong
+    if (r > 0) {
+        /* Invalidate the negative cache. */
+        if (impl->_abc_negative_cache != NULL &&
+                PySet_Clear(impl->_abc_negative_cache) < 0) {
+            goto end;
+        }
+        /* INCREF the new value of cache version,
+           then carefully DECREF the old one. */
+        Py_INCREF(abc_invalidation_counter);
+        Py_SETREF(impl->_abc_negative_cache_version, abc_invalidation_counter);
+    }
+    else {
+        incache = _in_weak_set(impl->_abc_negative_cache, subclass);
+        if (incache < 0) {
+            goto end;
+        }
+        if (incache > 0) {
+            result = Py_False;
+            goto end;
+        }
+    }
+
+    /* 3. Check the subclass hook. */
+    ok = _PyObject_CallMethodIdObjArgs((PyObject *)self, &PyId___subclasshook__,
+                                       subclass, NULL);
+    if (ok == NULL) {
+        goto end;
+    }
+    if (ok == Py_True) {
+        Py_DECREF(ok);
+        if (_add_to_weak_set(&impl->_abc_cache, subclass) < 0) {
+            goto end;
+        }
+        result = Py_True;
+        goto end;
+    }
+    if (ok == Py_False) {
+        Py_DECREF(ok);
+        if (_add_to_weak_set(&impl->_abc_negative_cache, subclass) < 0) {
+            goto end;
+        }
+        result = Py_False;
+        goto end;
+    }
+    if (ok != Py_NotImplemented) {
+        Py_DECREF(ok);
+        PyErr_SetString(PyExc_AssertionError, "__subclasshook__ must return either"
+                                              " False, True, or NotImplemented");
+        goto end;
+    }
+    Py_DECREF(ok);
+
+    /* 4. Check if it's a direct subclass. */
+    mro = ((PyTypeObject *)subclass)->tp_mro;
+    assert(PyTuple_Check(mro));
+    for (pos = 0; pos < PyTuple_Size(mro); pos++) {
+        PyObject *mro_item = PyTuple_GetItem(mro, pos);
+        if (mro_item == NULL) {
+            goto end;
+        }
+        if ((PyObject *)self == mro_item) {
+            if (_add_to_weak_set(&impl->_abc_cache, subclass) < 0) {
+                goto end;
+            }
+            result = Py_True;
+            goto end;
+        }
+    }
+
+    /* 5. Check if it's a subclass of a registered class (recursive). */
+    if (subclasscheck_check_registry(impl, subclass, &result)) {
+        // Exception occured or result is set.
+        goto end;
+    }
+
+    /* 6. Check if it's a subclass of a subclass (recursive). */
+    subclasses = PyObject_CallMethod(self, "__subclasses__", NULL);
+    if(!PyList_Check(subclasses)) {
+        PyErr_SetString(PyExc_TypeError, "__subclasses__() must return a list");
+        goto end;
+    }
+    for (pos = 0; pos < PyList_GET_SIZE(subclasses); pos++) {
+        int r = PyObject_IsSubclass(subclass, PyList_GET_ITEM(subclasses, pos));
```

Comment:
`PyList_GET_ITEM(subclasses, pos)` is a borrowed reference while `subclasses` is mutable and can has external references. It is safe to temporary increment the refcount.

Reply:
Hm, but the list itself holds strong references to all its elements, and will not be destroyed since we hold a strong reference to the list. Anyway, if you think it is important I can add an `INCREF`.

Reply:
`PyObject_IsSubclass()` can execute arbitrary Python code. This code can modify the list.

Reply:
OK, but note that this can happen only if a subclass overrides `__subclasses__`, because `object.__subclasses__` returns a new list on each call.